Add feature blog for CEL for Admission Control (#38162)
kubernetes/website: content/en/blog/_posts/2022-11-29-validating-admission-policies-alpha/index.md

monitored and have a well defined upgrade and rollback plan. To make matters
worse, if a webhook times out or is unavailable, the Kubernetes control plan can
become unavailable. This enhancement eliminates much of this complexity of
admission webhooks by embedding logic into the Kubernetes resources using CEL
instead of calling out to a remote webhook binary.

For example, to set a limit on deployment replicas, declare a policy:

```yaml
apiVersion: admissionregistration.k8s.io/v1alpha1
kind: ValidatingAdmissionPolicy
metadata:
  name: "demo-policy.example.com"
spec:
  matchConstraints:
    resourceRules:
    - apiGroups:   ["apps"]
      apiVersions: ["v1"]
      operations:  ["CREATE", "UPDATE"]
      resources:   ["deployments"]
  validations:
    - expression: "object.spec.replicas <= 5"
```

Note that the `expression` field contains a CEL expression that will be
evaluated by all resources matching the `matchConstraints`.

Next bind the policy the the appropriate resources:

```yaml
apiVersion: admissionregistration.k8s.io/v1alpha1
kind: ValidatingAdmissionPolicyBinding
metadata:
  name: "demo-binding-test.example.com"
spec:

>   policy: "replicalimit-policy.example.com"

Author: Good catch. My copy-paste got sloppy.. Fixing.

  matchResources:
    namespaceSelectors:
    - key: environment,
      operator: In,
      values: ["test"]
```

This `ValidatingAdmissionPolicyBinding` resource binds the policy we just
defined only to for namespaces where the `environment` label is set to `test`.

Compared to an admission webhook, admission policies are far simpler. The entire
policy is declared in Kubernetes resources. Unlike webhooks, there are no
external binaries or remote requests involved.

Validation admission policies are highly configurable, enabling policy authors
to define policies that can be parameterized and scoped to resources as needed
by cluster administrators.

For example, the above admission policy can be modified to make it configurable:

```yaml
apiVersion: admissionregistration.k8s.io/v1alpha1
kind: ValidatingAdmissionPolicy
metadata:
  name: "replicalimit-policy.example.com"
Spec:
  paramKind:

>     apiVersion: rules.example.com/v1

Reviewer: ```suggestion
    apiVersion: rules.example.com/v1 # You also need a CustomResourceDefinition for this API
```
?

Author: Thanks, this helps.

    kind: ReplicaLimit
  matchConstraints:
    resourceRules:
    - apiGroups:   ["apps"]
      apiVersions: ["v1"]
      operations:  ["CREATE", "UPDATE"]
      resources:   ["deployments"]
  validations:
    - expression: "object.spec.replicas <= params.maxReplicas"
```

Here, `paramKind` defines the resources used to configure the policy and the
`expression` uses the `params` variable to access the parameter resource.

This allows multiple bindings to be defined, each configured differently. For
example:

```yaml
apiVersion: admissionregistration.k8s.io/v1alpha1
kind: ValidatingAdmissionPolicyBinding
metadata:
  name: "replicalimit-binding-production.example.com"
spec:
  policy: "replicalimit-policy.example.com"
  paramsRef:
    name: "replica-limit-production.example.com"
  matchResources:
    namespaceSelectors:
    - key: environment,
      operator: In,
      values: ["production"]
```

```yaml

> apiVersion: rules.example.com/v1

Reviewer: ```suggestion
apiVersion: rules.example.com/v1 # defined via a CustomResourceDefinition
```
?

kind: ReplicaLimit
metadata:
  name: "replica-limit-production.example.com"
maxReplicas: 1000
```

This binding and parameter resource configure deployments in namespaces with the
`environment` label set to `production` to a max limit of 1000 replicas.

> A seperate binding can then be configured to set a lower limit for `test`
> namespaces.

Reviewer: Two nits:

---

“separate” (two `e`s, two `a`s)

---

```suggestion
You can then configure a separate binding in order to set a lower limit for namespaces
in the `test` environment.
```

I hope this has given you a glimpse of what is possible with validating
admission policies! There are many features that we have not yet touched on.

> Have a look at the [feature
> documentation](content/en/docs/reference/access-authn-authz/validating-admission-policy.md)
> to learn more.

Reviewer: ```suggestion
To learn more, read
[FIXME](/docs/reference/access-authn-authz/validating-admission-policy/).
```

Use the title of the page there. However, https://deploy-preview-37810--kubernetes-io-main-staging.netlify.app/docs/reference/access-authn-authz/validating-admission-policy/ doesn't work yet. Is there a PR that needs to land first?

Author: Correct, this requires https://github.com/kubernetes/website/pull/37770

We are working hard to add more features to admission policies and make the
enhancement easier to use. Try it out, send us your feedback and help us build
a simpler alternative to admission webhooks!

## How do I get involved?

If you want to get involved in development of admission policies, discuss enhancement
roadmaps, or report a bug, you can get in touch with developers at

> [SIG-API-Machinery](https://github.com/kubernetes/community/tree/master/sig-api-machinery).

Reviewer: ```suggestion
[SIG API Machinery](https://github.com/kubernetes/community/tree/master/sig-api-machinery).
```
